File: src/erlang/montecarlopi.erl
-module(montecarlopi).
-export([start/0, montecarlopi/4]).

montecarlopi(Iterations, Count, InCircle, Master) ->
  if
    Count /= Iterations ->
      X = rand:uniform(),
      Y = rand:uniform(),
      Length = X * X + Y * Y,
      if
        Length =< 1.0 ->
          montecarlopi(Iterations, Count + 1, InCircle + 1, Master);
        true ->
          montecarlopi(Iterations, Count + 1, InCircle, Master)
      end;
    true ->
      Master ! (4.0 * InCircle) / Iterations
  end.

gather(Count, Pi) ->
  if
    Count > 0 ->
      receive N -> gather(Count - 1, Pi + N) end;
    true ->
      Pi
  end.

run(Count, Iterations, Master) ->
  if
    Count > 0 ->
      spawn(montecarlopi, montecarlopi, [Iterations, 0, 0, Master]),
      run(Count - 1, Iterations, Master);
    true ->
      ok
  end.

save_results([], File) ->
  ok;
save_results([X|Results], File) ->
  file:write(File, io_lib:fwrite("~p\n", [X])),
  save_results(Results, File).

experiment(Count, Iterations, Threads, Results) ->
  if
    Count > 0 ->
      run(Threads, Iterations / Threads, self()),
      Pi = gather(Threads, 0.0) / Threads,
      {_, Time} = statistics(wall_clock),
      io:format("~p ~p ~p~n", [Count, Pi, Time]),
      experiment(Count - 1, Iterations, Threads, [Time|Results]);
    true ->
      {ok, File} = file:open(io_lib:fwrite("mcp-erl-~p.csv", [Threads]), [write]),
      save_results(Results, File)
  end.

start() ->
  experiment(100, math:pow(2, 24), 1, []),
  experiment(100, math:pow(2, 24), 2, []),
  experiment(100, math:pow(2, 24), 4, []),
  experiment(100, math:pow(2, 24), 8, []),
  experiment(100, math:pow(2, 24), 16, []).

File: src/go/montecarlopi.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"os"
	"time"
)

func montecarlopi(iterations int, master chan float64) {
	s := rand.NewSource(time.Now().UnixNano())
	r := rand.New(s)
	in_circle := 0
	for count := 0; count < iterations; count++ {
		x := r.Float64()
		y := r.Float64()
		if x*x+y*y <= 1.0 {
			in_circle++
		}
	}
	master <- (4.0 * float64(in_circle)) / float64(iterations)
}

func experiment(iterations int, threads int) {
	f, err := os.Create(fmt.Sprintf("mcp-go-%d.csv", threads))
	if err != nil {
		fmt.Println("error creating file: ", err)
		return
	}
	c := make(chan float64)
	defer f.Close()
	for index := 0; index < 100; index++ {
		start := time.Now()
		for t := 0; t < threads; t++ {
			go montecarlopi(iterations/threads, c)
		}
		pi := 0.0
		for t := 0; t < threads; t++ {
			pi = pi + <-c
		}
		pi = pi / float64(threads)
		total := time.Now().Sub(start).Nanoseconds()
		fmt.Println(index, " ", pi, " ", total)
		_, err := f.WriteString(fmt.Sprintf("%d\n", total))
		if err != nil {
			fmt.Println("error writing string: ", err)
		}
	}
}

func main() {
	experiment(int(math.Pow(2, 24)), 1)
	experiment(int(math.Pow(2, 24)), 2)
	experiment(int(math.Pow(2, 24)), 4)
	experiment(int(math.Pow(2, 24)), 8)
	experiment(int(math.Pow(2, 24)), 16)
}
